capsules/src/usb/cdc.rs

```diff
@@ -0,0 +1,496 @@
+                // Nothing to do for CDC ACM.
+                hil::usb::InResult::Delay
+            }
+        }
+    }
+
+    /// Handle a Bulk/Interrupt OUT transaction
+    fn packet_out(
+        &'a self,
+        transfer_type: TransferType,
+        endpoint: usize,
+        packet_bytes: u32,
+    ) -> hil::usb::OutResult {
+        match transfer_type {
+            TransferType::Bulk => {
+                // Start by checking to see if we even care about this RX or
+                // not.
+                self.rx_buffer.take().map(|rx_buf| {
+                    let rx_offset = self.rx_offset.get();
+
+                    // How many more bytes can we store in our RX buffer?
+                    let available_bytes = rx_buf.len() - rx_offset;
+                    let copy_length = cmp::min(packet_bytes as usize, available_bytes);
+
+                    // Do the copy into the RX buffer.
+                    let packet = self.buffer(endpoint);
+                    for i in 0..copy_length {
+                        rx_buf[rx_offset + i] = packet[i].get();
+                    }
+
+                    // Keep track of how many bytes we have received so far.
+                    let total_received_bytes = rx_offset + copy_length;
+
+                    // Update how many bytes we have gotten.
+                    self.rx_offset.set(total_received_bytes);
+
+                    // Check if we have received at least as many bytes as the
+                    // client asked for.
+                    if total_received_bytes >= self.rx_len.get() {
+                        self.rx_client.map(move |client| {
+                            client.received_buffer(
+                                rx_buf,
+                                total_received_bytes,
+                                ReturnCode::SUCCESS,
+                                uart::Error::None,
+                            );
+                        });
+                    } else {
+                        // Make sure to put the RX buffer back.
+                        self.rx_buffer.replace(rx_buf);
+                    }
+                });
+
+                // No error cases to report to the USB.
+                hil::usb::OutResult::Ok
+            }
+            TransferType::Control | TransferType::Isochronous | TransferType::Interrupt => {
+                // Nothing to do for CDC ACM.
+                hil::usb::OutResult::Ok
+            }
+        }
+    }
+
+    fn packet_transmitted(&'a self, _endpoint: usize) {
+        // Check if more to send.
+        self.tx_buffer.take().map(|tx_buf| {
+            // Check if we have any bytes to send.
+            let remaining = self.tx_len.get() - self.tx_offset.get();
+            if remaining > 0 {
+                // We do, so ask to send again.
+                self.tx_buffer.replace(tx_buf);
+                self.controller().endpoint_resume_in(ENDPOINT_IN_NUM);
```

Comment:
Shouldn't we be checking if `ENDPOINT_IN_NUM == _endpoint` here?

Reply:
We only have one IN endpoint.

Reply:
Seems fine then.